demos/sample_azure_iot_pnp/sample_azure_iot_pnp.c

```diff
@@ -0,0 +1,1062 @@
+        {
+            LogError( ( "There was an error parsing the properties: 0x%08x", xResult ) );
+        }
+        else
+        {
+            LogInfo( ( "Successfully parsed properties" ) );
+            xResult = eAzureIoTSuccess;
+        }
+    }
+
+    return xResult;
+}
+/*-----------------------------------------------------------*/
+
+/**
+ * @brief Update local device temperature values based on new requested temperature.
+ */
+static void prvUpdateLocalProperties( double xNewTemperatureValue,
+                                      uint32_t ulPropertyVersion,
+                                      bool * pxOutMaxTempChanged )
+{
+    *pxOutMaxTempChanged = false;
+    xDeviceCurrentTemperature = xNewTemperatureValue;
+
+    /* Update maximum or minimum temperatures. */
+    if( xDeviceCurrentTemperature > xDeviceMaximumTemperature )
+    {
+        xDeviceMaximumTemperature = xDeviceCurrentTemperature;
+        *pxOutMaxTempChanged = true;
+    }
+    else if( xDeviceCurrentTemperature < xDeviceMinimumTemperature )
+    {
+        xDeviceMinimumTemperature = xDeviceCurrentTemperature;
+    }
+
+    /* Calculate the new average temperature. */
+    ulDeviceTemperatureCount++;
+    xDeviceTemperatureSummation += xDeviceCurrentTemperature;
+    xDeviceAverageTemperature = xDeviceTemperatureSummation / ulDeviceTemperatureCount;
+
+    LogDebug( ( "Client updated desired temperature variables locally." ) );
+    LogDebug( ( "Current Temperature: %2f", xDeviceCurrentTemperature ) );
+    LogDebug( ( "Maximum Temperature: %2f", xDeviceMaximumTemperature ) );
+    LogDebug( ( "Minimum Temperature: %2f", xDeviceMinimumTemperature ) );
+    LogDebug( ( "Average Temperature: %2f", xDeviceAverageTemperature ) );
+}
+/*-----------------------------------------------------------*/
+
+/**
+ * @brief Send updated maximum temperature value to IoT Hub.
+ */
+static void prvSendNewMaxTemp( double xUpdatedTemperature )
+{
+    AzureIoTResult_t xResult;
+    AzureIoTJSONWriter_t xWriter;
+    int32_t lBytesWritten;
+
+    xResult = AzureIoTJSONWriter_Init( &xWriter, ucPropertyPayloadBuffer, sizeof( ucPropertyPayloadBuffer ) );
+    configASSERT( xResult == eAzureIoTSuccess );
+
+    xResult = AzureIoTJSONWriter_AppendBeginObject( &xWriter );
+    configASSERT( xResult == eAzureIoTSuccess );
+
+    xResult = AzureIoTJSONWriter_AppendPropertyName( &xWriter, sampleazureiotPROPERTY_MAX_TEMPERATURE_TEXT,
+                                                     strlen( sampleazureiotPROPERTY_MAX_TEMPERATURE_TEXT ) );
+    configASSERT( xResult == eAzureIoTSuccess );
+
+    xResult = AzureIoTJSONWriter_AppendDouble( &xWriter, xUpdatedTemperature, sampleazureiotDOUBLE_DECIMAL_PLACE_DIGITS );
+    configASSERT( xResult == eAzureIoTSuccess );
+
+    xResult = AzureIoTJSONWriter_AppendEndObject( &xWriter );
+    configASSERT( xResult == eAzureIoTSuccess );
+
+    lBytesWritten = AzureIoTJSONWriter_GetBytesUsed( &xWriter );
+
+    if( lBytesWritten < 0 )
```

Comment:
its just weird to see configAssert for all previous JSON errors, can we be consistent?

Reply:
We can. My thinking here is most of those JSON API's are relatively static. Buffer is the same size, payload is predictable, there isn't a whole lot to go wrong. For the send side it can be a little more complicated.

Do you mean for the getBytesUsed api only?

Reply:
AzureIoTJSONWriter_GetBytesUsed

Reply:
that's fair. I'm going to take a simplifying assumption that the json builder is going to succeed and use config asserts here

Reply:
this change is missing